lib/compat/wordpress-6.7/class-gutenberg-rest-templates-controller-6-7.php

```diff
@@ -0,0 +1,203 @@
+			if ( ! empty( $links['self']['href'] ) ) {
+				$actions = $this->get_available_actions();
+				$self    = $links['self']['href'];
+				foreach ( $actions as $rel ) {
+					$response->add_link( $rel, $self );
+				}
+			}
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Returns the source from where the template originally comes from.
+	 *
+	 * @param WP_Block_Template $template_object Template instance.
+	 * @return string                            Original source of the template one of theme, plugin, site, or user.
+	 */
+	// @core-merge: Changed the comments format (from inline to multi-line) in the entire function.
+	private static function get_wp_templates_original_source_field( $template_object ) {
+		if ( 'wp_template' === $template_object->type || 'wp_template_part' === $template_object->type ) {
+			/*
+			 * Added by theme.
+			 * Template originally provided by a theme, but customized by a user.
+			 * Templates originally didn't have the 'origin' field so identify
+			 * older customized templates by checking for no origin and a 'theme'
+			 * or 'custom' source.
+			 */
+			if ( $template_object->has_theme_file &&
+			( 'theme' === $template_object->origin || (
+				empty( $template_object->origin ) && in_array(
+					$template_object->source,
+					array(
+						'theme',
+						'custom',
+					),
+					true
+				) )
+			)
+			) {
+				return 'theme';
+			}
+
+			// Added by plugin.
+			// @core-merge: Removed `$template_object->has_theme_file` check from this if clause.
+			if ( 'plugin' === $template_object->origin ) {
+				return 'plugin';
+			}
+
+			/*
+			 * Added by site.
+			 * Template was created from scratch, but has no author. Author support
+			 * was only added to templates in WordPress 5.9. Fallback to showing the
+			 * site logo and title.
+			 */
+			if ( empty( $template_object->has_theme_file ) && 'custom' === $template_object->source && empty( $template_object->author ) ) {
+				return 'site';
+			}
+		}
+
+		// Added by user.
+		return 'user';
+	}
+
+	/**
+	 * Returns a human readable text for the author of the template.
+	 *
+	 * @param WP_Block_Template $template_object Template instance.
+	 * @return string                            Human readable text for the author.
+	 */
+	private static function get_wp_templates_author_text_field( $template_object ) {
+		$original_source = self::get_wp_templates_original_source_field( $template_object );
+		switch ( $original_source ) {
+			case 'theme':
+				$theme_name = wp_get_theme( $template_object->theme )->get( 'Name' );
+				return empty( $theme_name ) ? $template_object->theme : $theme_name;
+			case 'plugin':
+				// @core-merge: Prioritize plugin name instead of theme name for plugin-registered templates.
+				if ( ! function_exists( 'get_plugins' ) || ! function_exists( 'get_plugin_data' ) ) {
+					require_once ABSPATH . 'wp-admin/includes/plugin.php';
+				}
+				if ( isset( $template_object->plugin ) ) {
+					$plugins = wp_get_active_and_valid_plugins();
```

Comment:
```suggestion
					$plugins = get_plugins();
```

No need to have the active and valid check. We just need to find the plugin data.

Reply:
@TimothyBJacobs This was an earlier suggestion of mine. `get_plugins()` scans all the files in the plugin directory so is quite expensive. `wp_get_active_and_valid_plugins()` uses data from the options table so is significantly faster.

As the registered templates will be from active plugins I think it's fine to use the faster function but there's a chance I am missing something. Am I?

Reply:
My thinking is `get_plugins` caches after the first time through, and will also pull the plugin data for us which we are going to need. It also looks like there already is a call to `get_plugins`? I'm not entirely sure what code path that's for.

I could definitely be wrong there, though. It's probably worth pulling into a method and benching it.

Reply:
> It also looks like there already is a call to get_plugins? I'm not entirely sure what code path that's for.

That's basically to "fall back" to the current behavior in `trunk`:

https://github.com/WordPress/wordpress-develop/blob/9f09c574c0ac64041c81760f10646994d21ce49d/src/wp-includes/rest-api/endpoints/class-wp-rest-templates-controller.php#L868

I could refactor it to use `wp_get_active_and_valid_plugins()` if needed. I don't have a strong opinion between `get_plugins()` and `wp_get_active_and_valid_plugins()`, though.

Reply:
@TimothyBJacobs is this a blocker for this PR to land? It doesn't seem a big change to commit your suggestion and in general I think we would be usually at the `active` plugin use case, so maybe `wp_get_active_and_valid_plugins` is a bit lighter?